prefetch.bs

```diff
@@ -396,3 +396,16 @@ The <dfn>list of sufficiently strict speculative navigation referrer policies</d
 
     <div class="issue">Further review is needed of the cookies-present handling and its interaction with redirects and caches.</div>
 </div>
+
+<div algorithm>
+    To <dfn export>prefetch</dfn> given a {{Document}} |document|, [=URL=] |url| and [=referrer policy=] |referrerPolicy|, perform the following steps.
+
+    1. Let |partitionKey| be the result of [=determining the network partition key=] given |document|'s [=relevant settings object=].
+    1. Let |topLevelOrigin| be |document|'s [=relevant settings object=]'s [=environment/top-level origin=] if |document|'s [=Document/browsing context=] is not a [=top-level browsing context=], and |url|'s [=origin=] otherwise.
+    1. Let |topLevelSite| be the result of [=obtaining a site=], given |topLevelOrigin|.
+    1. Let |secondKey| be null or an [=implementation-defined=] value.
+    1. If |partitionKey| is equal to (|topLevelSite|, |secondKey|), then [=partitioned prefetch=] given |document|, |url| and |referrerPolicy|.
+    1. Otherwise, [=uncredentialed prefetch=] given |document|, |url| and |referrerPolicy|.
+
+    <div class="note">This determines whether the navigation would use the same network partition key. If it would, the prefetch is restricted to the same partition. Otherwise it is uncredentialed.</div>
```

Comment:
So in the redirect case the resulting partition key might not end up the same, since |url| is not the same as the final URL. Is that an issue?

Reply:
This is certainly an awkward case. At the moment the Chromium new-prefetch path doesn't allow this at all; what I've suggested in spec text is the following resolution (open to better ones):

- if it starts with a URL in the same partition, the "partitioned prefetch" algorithm is used which allows following redirects as long as they stay within the partition, and bails otherwise (at the top level, this basically means same-site)
- if it starts with a URL in another partition, the "uncredentialed prefetch" algorithm is used which allows redirects as long as they remain outside the requesting page's partition (i.e., they remain cross-site) and isolates the storage

This obviously doesn't deal with redirect chains which include both same- and cross-site URLs, but it doesn't preclude adding specific cases of that which are safe at a later point.

Reply:
Oh, that sounds great. Maybe just add that explanation in a note here, since it's pretty easy to say "wait, |url| doesn't look like the right way to choose a partition key", without realizing that if you follow all the algorithms you'll end up in a safe place.